parkbenchcommon/broadcaster.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2017-2018 Joel Allen Luellwitz and Andrew Klapp
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Provides the broadcasting component of a filesystem-based IPC mechanism."""
+
+__all__ = ['BroadcasterBroadcastException',
+           'BroadcasterInitException',
+           'Broadcaster']
+
+import datetime
+import logging
+import os
+import stat
+import traceback
+import tmpfs
+
+TMPFS_SIZE = '1M'
+
+class BroadcasterBroadcastException(Exception):
+    """This exception is raised when a BeaconBroadcaster object fails to issue a broadcast."""
+
+class BroadcasterInitException(Exception):
+    """This exception is raised when a BeaconBroadcaster object fails to initialize."""
+
+class Broadcaster(object):
+    """Provides the broadcasting component of a filesystem-based IPC mechanism."""
+
+    def __init__(self, program_name, broadcast_name, uid, gid):
+        """Initial configuration of the broadcast directory. This must be done as root. This
+        method creates any necessary spool directories, sets their permissions, and mounts a
+        ramdisk if necessary.
+
+        program_name: The name of the program issuing the broadcast.
+        broadcast_name: The name of the broadcast to issue.
+        uid: The UID of the calling program.
+        gid: The GID of the calling program.
+        """
+
+        self.logger = logging.getLogger(__name__)
+
+        self.logger.debug("Initializing broadcaster for broadcast %s from program %s.",
+                          broadcast_name, program_name)
+
+
+        broadcast_path = '/var/spool/'
+        self.broadcast_path = os.path.join(broadcast_path, program_name, broadcast_name)
+        self.program_name = program_name
+        self.broadcast_name = broadcast_name
+
+        tmpfs_path = os.path.join(broadcast_path, program_name)
+        tmpfs.mount_tmpfs(tmpfs_path, TMPFS_SIZE)
+
+        self._create_directory(self.broadcast_path)
```

Comment:
Again, permissions should be set when you create the directory in the same manner as I did with the proper-daemonizing stuff. As I said below, they might also have to be set when we mount the ramdisk. Maybe creating the directory (directories?) and mounting the ramdisk should be combined somehow. Alternatively, maybe mounting the ramdisk shouldn't create any directories at all?